In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
data = pd.read_csv('AQI_Data.csv')

In [3]:
data.head(n=50)

city  AQI PM2.5 PM10 SO2 NO2   CO  O3 main_pollution                 time
0     北京市   61    14   71   3  12  0.3  61           PM10  2019-02-04 00:00:00
1     天津市  104    41  157   8  21  0.9  53           PM10  2019-02-04 00:00:00
2    石家庄市  125    36  200  12  47  0.9  30           PM10  2019-02-04 00:00:00
3     唐山市   66    42   81  22  16  1.0  52           PM10  2019-02-04 00:00:00
4    秦皇岛市   55    25   60  12  22  0.9  48           PM10  2019-02-04 00:00:00
5     邯郸市  240   190  329  31  62  2.5  14          PM2.5  2019-02-04 00:00:00
6     邢台市  254   204  300  26  61  2.2  18          PM2.5  2019-02-04 00:00:00
7     保定市  102    42  153   6  18  0.7  54           PM10  2019-02-04 00:00:00
8     承德市   61    19   72   7  19  0.8  39           PM10  2019-02-04 00:00:00
9     沧州市  109    43  167  23  19  1.0  51           PM10  2019-02-04 00:00:00
10    廊坊市   68    22   85   4  16  0.5  56           PM10  2019-02-04 00:00:00
11    衡水市  190   143  227  24  30  1.6  56          PM2.5  2019-02-04 00:00:00
12   张家口市   58    22   66   9   9  0.6  63           PM10  2019-02-04 00:00:00
13    太原市   92    37  134  17  28  0.7  50           PM10  2019-02-04 00:00:00
14    大同市  103    30  156  17  10  0.5  70           PM10  2019-02-04 00:00:00
15    阳泉市   68    33   85  12  13  0.5  67           PM10  2019-02-04 00:00:00
16    长治市   60    30   70  26  36  1.0  49           PM10  2019-02-04 00:00:00
17    晋城市   62    14   73  12  13  0.5  76           PM10  2019-02-04 00:00:00
18    朔州市   85    18  119  10   7  0.4  77           PM10  2019-02-04 00:00:00
19    晋中市   90    59  130  72  40  1.0  30           PM10  2019-02-04 00:00:00
20    运城市  104    63  157  34  53  1.3  27           PM10  2019-02-04 00:00:00
21    忻州市   69    20   87  20  12  0.5  75           PM10  2019-02-04 00:00:00
22    临汾市   77    56  102  60  38  1.4  30          PM2.5  2019-02-04 00:00:00
23    吕梁市   89    49  127  76  26  0.8  38           PM10  2019-02-04 00:00:00
24  呼和浩特市   96    31  142  16  19  0.8  54           PM10  2019-02-04 00:00:00
25    包头市   86    44  122  28  49  1.4  29           PM10  2019-02-04 00:00:00
26    乌海市   77    35  104  28  20  1.0  56           PM10  2019-02-04 00:00:00
27    赤峰市   49    24   49  20  12  0.5  52              —  2019-02-04 00:00:00
28    通辽市   41    27   41   9   6  0.2  59              —  2019-02-04 00:00:00
29  鄂尔多斯市   80    22  110   9  12  0.6  69           PM10  2019-02-04 00:00:00
30  呼伦贝尔市   35    24   32   4   5  0.0  61              —  2019-02-04 00:00:00
31  巴彦淖尔市  100    32  150  10   8  0.6  64           PM10  2019-02-04 00:00:00
32  乌兰察布市   82    24  114  19   6  0.3  85           PM10  2019-02-04 00:00:00
33    兴安盟   30    21   27   6   7  0.3  58              —  2019-02-04 00:00:00
34  锡林郭勒盟   16     7   16   8   2  0.2  49              —  2019-02-04 00:00:00
35   阿拉善盟   63    32   76   8   6  0.5  88           PM10  2019-02-04 00:00:00
36    沈阳市   57    22   64  12  14  0.5  55           PM10  2019-02-04 00:00:00
37    大连市   58    23   66  10  10  0.6  63           PM10  2019-02-04 00:00:00
38   瓦房店市   62    19   74  18  20  0.7  60           PM10  2019-02-04 00:00:00
39    鞍山市   71    34   91  27  25  1.2  47           PM10  2019-02-04 00:00:00
40    抚顺市   55     9   60  10   6  0.4  67           PM10  2019-02-04 00:00:00
41    本溪市   59    19   68  12  10  0.8  56           PM10  2019-02-04 00:00:00
42    丹东市   61    32   72  12   8  0.5  62           PM10  2019-02-04 00:00:00
43    锦州市   32    16   32  31  16  0.8  55              —  2019-02-04 00:00:00
44    营口市   57    23   64  10  13  0.8  51           PM10  2019-02-04 00:00:00
45    阜新市   40    20   40  14   9  0.5  53              —  2019-02-04 00:00:00
46    辽阳市   68    28   86  10  16  0.6  47           PM10  2019-02-04 00:00:00
47    盘锦市   38    14   38  11  10  0.9  57              —  2019-02-04 00:00:00
48    铁岭市   44    12   44   5   8  0.3  62              —  2019-02-04 00:00:00
49    朝阳市   55    38   60  31  22  1.2  36           PM

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71825 entries, 0 to 71824
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   city            71825 non-null  object
 1   AQI             71825 non-null  object
 2   PM2.5           71825 non-null  object
 3   PM10            71825 non-null  object
 4   SO2             71825 non-null  object
 5   NO2             71825 non-null  object
 6   CO              71825 non-null  object
 7   O3              71825 non-null  object
 8   main_pollution  71825 non-null  object
 9   time            71825 non-null  object
dtypes: object(10)
memory usage: 5.5+ MB


In [5]:
data['time'].unique()

array(['2019-02-04 00:00:00', '2019-02-04 01:00:00',
       '2019-02-04 02:00:00', '2019-02-04 03:00:00',
       '2019-02-04 04:00:00', '2019-02-04 05:00:00',
       '2019-02-04 06:00:00', '2019-02-04 07:00:00',
       '2019-02-04 08:00:00', '2019-02-04 09:00:00',
       '2019-02-04 10:00:00', '2019-02-04 11:00:00',
       '2019-02-04 12:00:00', '2019-02-04 13:00:00',
       '2019-02-04 14:00:00', '2019-02-04 15:00:00',
       '2019-02-04 16:00:00', '2019-02-04 17:00:00',
       '2019-02-04 18:00:00', '2019-02-04 19:00:00',
       '2019-02-04 20:00:00', '2019-02-04 21:00:00',
       '2019-02-04 22:00:00', '2019-02-04 23:00:00',
       '2019-02-05 00:00:00', '2019-02-05 01:00:00',
       '2019-02-05 02:00:00', '2019-02-05 03:00:00',
       '2019-02-05 04:00:00', '2019-02-05 05:00:00',
       '2019-02-05 06:00:00', '2019-02-05 07:00:00',
       '2019-02-05 08:00:00', '2019-02-05 09:00:00',
       '2019-02-05 10:00:00', '2019-02-05 11:00:00',
       '2019-02-05 12:00:00', '2019-02-05 13:0

In [6]:
# 数据清洗
data['time'] = pd.to_datetime(data['time'])
data = data[data['time']<=pd.to_datetime('2019-02-11 23:59:59')]  # 选取2月4日——2月12日的数据
data.set_index(data["city"], inplace=False)
data = data.replace('—', np.nan)
for col in ['AQI', 'PM2.5', 'PM10', 'SO2', 'NO2', 'CO', 'O3']:
    data[col] = data[col].astype(float)
data.head()

city    AQI  PM2.5   PM10   SO2   NO2   CO    O3 main_pollution       time
0   北京市   61.0   14.0   71.0   3.0  12.0  0.3  61.0           PM10 2019-02-04
1   天津市  104.0   41.0  157.0   8.0  21.0  0.9  53.0           PM10 2019-02-04
2  石家庄市  125.0   36.0  200.0  12.0  47.0  0.9  30.0           PM10 2019-02-04
3   唐山市   66.0   42.0   81.0  22.0  16.0  1.0  52.0           PM10 2019-02-04
4  秦皇岛市   55.0   25.0   60.0  12.0  22.0  0.9  48.0           PM10 2019-02-04

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68155 entries, 0 to 68154
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   city            68155 non-null  object        
 1   AQI             67903 non-null  float64       
 2   PM2.5           67815 non-null  float64       
 3   PM10            66437 non-null  float64       
 4   SO2             67832 non-null  float64       
 5   NO2             67852 non-null  float64       
 6   CO              67838 non-null  float64       
 7   O3              67842 non-null  float64       
 8   main_pollution  41176 non-null  object        
 9   time            68155 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(7), object(2)
memory usage: 5.7+ MB


In [8]:
len(data['city'].unique())

367

In [9]:
data.to_csv('data_clean.csv', index=False)

#### 一、数据详情
- 共有68155条数据，时间为2月4日到2月11日，每隔1小时收集一次，囊括全国367个城市
- 包含字段：城市、空气质量指数、PM2.5指数、PM10指数、SO2指数、NO2指数、CO指数、O3指数、主要污染源、时间

#### 二、问题
1. 燃放烟花爆竹真的对空气质量有影响吗？
2. 烟花爆竹对空气质量的影响体现在哪些指标上？
3. 烟花爆竹对空气质量的污染程度有多大？
4. 哪些区域污染最严重(轻)？
5. 哪些城市属于一秒破功型？
6. 除夕中午到初一中午的超标城市个数
7. 除夕到初七的超标城市个数


##### 1. 燃放烟花爆竹真的对空气质量有影响吗？

In [10]:
AQI_total_mean = data[data['time']<=pd.to_datetime('2019-02-06 23:59:59')].groupby(['time'])['AQI'].mean()

In [11]:
from pyecharts import Line 

line = Line("全国春节期间空气质量指数总体趋势", "2019年除夕到初二", width=800)
line.add("", AQI_total_mean.index, np.round(AQI_total_mean.values,0), is_smooth=True, 
         legend_text_size=18,xaxis_label_textsize=14,yaxis_label_textsize=18,
         xaxis_rotate=20, yaxis_min=8, mark_point=["max"])
line

In [12]:
def city_AQI(data, cities):
    line = Line("春节期间空气质量指数", "2019年除夕到初二", width=800)
    for city in cities:
        city_aqi = data[(data['city']==city)&(data['time']<=pd.to_datetime('2019-02-06 23:59:59'))]
        line.add(city, city_aqi.index, np.round(city_aqi['AQI'].values,0), is_smooth=True, 
             legend_text_size=18,xaxis_label_textsize=14,yaxis_label_textsize=18,
             xaxis_rotate=20, yaxis_min=8, mark_point=["max"])
    return line

In [13]:
city_AQI(data=data, cities=['北京市', '天津市'])

In [14]:
city_AQI(data=data, cities=['长沙市', '南宁市'])

##### 2. 烟花爆竹对空气质量的影响体现在哪些指标上？

In [15]:
data_total_idx = data[data['time']<=pd.to_datetime('2019-02-06 23:59:59')].groupby(['time']).mean()

In [16]:
data_total_idx.columns

Index(['AQI', 'PM2.5', 'PM10', 'SO2', 'NO2', 'CO', 'O3'], dtype='object')

In [17]:
line = Line("全国春节期间空气质量各指标（2019年除夕到初二）", "单位:μg/m3(CO为mg/m3)", width=800)
for idx in data_total_idx.columns[1:]:
    line.add(idx, data_total_idx.index, np.round(data_total_idx[idx].values,0), is_smooth=True, 
             legend_text_size=18,xaxis_label_textsize=14,yaxis_label_textsize=18,
             xaxis_rotate=20, yaxis_min=8, legend_top=30)
line

In [18]:
tianjin_idx = data[(data['city']=='石家庄市')&(data['time']<=pd.to_datetime('2019-02-06 23:59:59'))]
line = Line("石家庄市春节期间空气质量各指标（2019年除夕到初二）", "单位:μg/m3(CO为mg/m3)", width=800)
for idx in tianjin_idx.columns[2:8]:
    line.add(idx, tianjin_idx.index, np.round(tianjin_idx[idx].values,0), is_smooth=True, 
             legend_text_size=18,xaxis_label_textsize=14,yaxis_label_textsize=18,
             xaxis_rotate=20, yaxis_min=8, legend_top=30)
line

In [19]:
main_pol = data.loc[data['time']==pd.to_datetime('2019-02-05 02:00:00'), 'main_pollution'].value_counts()

In [20]:
main_pol

PM2.5         243
PM10           97
PM10,PM2.5     10
Name: main_pollution, dtype: int64

In [21]:
from pyecharts import Bar

bar = Bar("全国城市首要污染物", "2019年2月5日凌晨2点", width=600)
bar.add("", main_pol.index, main_pol.values, is_stack=True, is_label_show=True,
       bar_category_gap='40%', label_color = ['#130f40'], label_text_size=18,
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

##### 3. 烟花爆竹对空气质量的污染程度有多大？

In [22]:
data2 = data[data['time']<=pd.to_datetime('2019-02-05 23:59:59')]
data_AQI_min = data2.groupby('city')['AQI'].min()
data_AQI_max = data2.groupby('city')['AQI'].max()

In [23]:
data_AQI_times = np.round(data_AQI_max/data_AQI_min, 1)

In [24]:
data_AQI_times_top10 = data_AQI_times.nlargest(10)

In [25]:
bar = Bar("全国除夕和春节期间空气质量最高最低比Top10城市", "时间：2019年除夕至初一", width=600)
bar.add("", data_AQI_times_top10.index, data_AQI_times_top10.values, is_stack=True, 
        is_label_show=True,bar_category_gap='40%', label_color = ['#130f40'], label_text_size=18,
       legend_text_size=18,xaxis_label_textsize=14,yaxis_label_textsize=18, xaxis_rotate=30)
bar

In [26]:
data_AQI_times_top10

city
伊春市       22.7
鸡西市       22.7
锡林郭勒盟     21.4
大兴安岭地区    20.4
玉树州       19.2
盘锦市       17.9
葫芦岛市      17.2
甘南州       16.7
瓦房店市      16.1
锦州市       16.1
Name: AQI, dtype: float64

In [27]:
city_AQI(data=data, cities=['伊春市','鸡西市', '大兴安岭地区'])

In [28]:
city_AQI(data=data, cities=['盘锦市','葫芦岛市','瓦房店市','锦州市'])

In [29]:
data_AQI_times_counts = data_AQI_times.value_counts(bins=[1,5,10,15,20,25])
data_AQI_times_counts

(0.999, 5.0]    238
(5.0, 10.0]      88
(10.0, 15.0]     28
(15.0, 20.0]      9
(20.0, 25.0]      4
Name: AQI, dtype: int64

In [30]:
bar = Bar("全国除夕和春节期间空气质量最高最低比的城市数量", "时间：2019年除夕至初一", width=600)
times = ['1-5倍','5-10倍','10-15倍','15-20倍','20-25倍']
bar.add("", times, data_AQI_times_counts.values, is_stack=True, 
        is_label_show=True,bar_category_gap='40%', label_color = ['#130f40'], label_text_size=18,
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

##### 4. 哪些区域污染最严重（轻）？

In [31]:
start_time = pd.to_datetime('2019-02-04 20:00:00')
end_time = pd.to_datetime('2019-02-05 04:00:00')
AQI_by_city = data[(data['time']>=start_time)&(data['time']<=end_time)].groupby('city')['AQI'].mean()

In [32]:
from pyecharts import Geo

geo = Geo(
    "全国各城市空气质量指数",
    "除夕晚上20点至初一凌晨4点平均值",
    title_color="#fff",
    title_pos="center",
    width=800,
    height=500,
    background_color="#404a59",
)
attr, value = AQI_by_city.index, AQI_by_city.values

geo.add(
    "",
    attr,
    value,
    visual_range=[0, 500],
    visual_text_color="#fff",
    symbol_size=10,
    is_visualmap=True,
    is_piecewise=True,
    visual_split_number=10,
)
geo

In [33]:
AQI_by_city.nlargest(10)

city
阿克苏地区    454.333333
库尔勒市     436.111111
呼和浩特市    394.333333
和田地区     357.666667
安康市      343.888889
白山市      340.333333
亳州市      320.333333
喀什地区     320.111111
包头市      295.666667
咸阳市      288.222222
Name: AQI, dtype: float64

In [34]:
city_AQI(data=data, cities=['呼和浩特市', '包头市', '咸阳市'])

In [35]:
AQI_by_city.nsmallest(5)

city
义乌市      12.500000
阿勒泰地区    27.111111
攀枝花市     27.444444
丽江市      28.000000
阿里地区     28.222222
Name: AQI, dtype: float64

In [36]:
AQI_by_city[AQI_by_city<=50]

city
中山市      41.555556
丽江市      28.000000
义乌市      12.500000
保山市      42.888889
嘉峪关市     34.888889
塔城地区     33.666667
大理州      41.444444
安顺市      39.333333
惠州市      47.333333
拉萨市      36.777778
揭阳市      45.222222
攀枝花市     27.444444
昆明市      48.666667
景德镇市     43.444444
林芝市      36.333333
武威市      47.777778
汕尾市      49.555556
江门市      42.555556
海北州      39.111111
海南州      44.333333
海口市      41.333333
海西州      46.333333
深圳市      44.777778
珠海市      48.777778
甘孜州      43.555556
迪庆州      39.111111
那曲地区     41.333333
金昌市      49.000000
阳江市      38.444444
阿勒泰地区    27.111111
阿坝州      41.222222
阿里地区     28.222222
Name: AQI, dtype: float64

##### 5. 哪些城市属于一秒破功型？

In [37]:
city1 = data.loc[(data['time'] == pd.to_datetime('2019-02-05 02:00:00'))&(data['AQI']<100), 'city']

In [38]:
data3 = data[(data['time'] >= pd.to_datetime('2019-02-05 18:00:00'))&
            (data['time'] <= pd.to_datetime('2019-02-06 23:59:59'))]

In [39]:
city2 = data3.loc[data3['AQI']>200, 'city']

In [40]:
list(set(city1).intersection(set(city2)))

['长春市', '中卫市', '那曲地区', '铜川市', '忻州市', '白城市', '郑州市', '吴忠市', '晋城市']

In [41]:
city_AQI(data=data, cities=['郑州市', '长春市', '忻州市', '晋城市'])

##### 6. 除夕中午到初一中午的超标城市个数

In [42]:
start_time = pd.to_datetime('2019-02-04 12:00:00')
end_time = pd.to_datetime('2019-02-05 12:00:00')

In [43]:
city_over_new_year = data.loc[(data['time']>=start_time)&(data['time']<=end_time)&(data['AQI']>100)].groupby('time')['AQI'].count()

In [44]:
from pyecharts import Overlap

bar = Bar("全国除夕中午至初一中午空气质量超标城市数量", "时间: 2019年  总城市数: 367", width=1000)
new_year_time = ['除夕12点','除夕13点','除夕14点','除夕15点','除夕16点','除夕17点','除夕18点','除夕19点',
        '除夕20点','除夕21点','除夕22点','除夕23点','春节0点','春节1点','春节2点','春节3点',
        '春节4点','春节5点','春节6点','春节7点','春节8点','春节9点','春节10点','春节11点','春节12点']
bar.add("", new_year_time, city_over_new_year.values, is_stack=True, 
        is_label_show=True,bar_category_gap='40%', label_color = ['#130f40'], label_text_size=18,
       legend_text_size=18,xaxis_label_textsize=14,yaxis_label_textsize=18, xaxis_rotate=90)

line = Line("", width=1000)
line.add("", new_year_time, city_over_new_year.values+5, is_smooth=True)

overlap = Overlap(width=1000)
overlap.add(bar)
overlap.add(line)

overlap

##### 7. 除夕到初七的超标城市个数

In [ ]:
city_over_count = (data.groupby('city')['AQI'].resample('D').mean()>100).unstack(level=-1).sum()
city_over_count

In [46]:
bar = Bar("全国除夕至初七期间空气质量超标城市数量", "时间: 2019年  总城市数: 367", width=800)
date = ['02-04(除夕)','02-05(初一)','02-06(初二)','02-07(初三)',
        '02-08(初四)','02-09(初五)','02-10(初六)','02-11(初七)']
bar.add("", date, city_over_count.values, is_stack=True, 
        is_label_show=True,bar_category_gap='40%', label_color = ['#130f40'], label_text_size=18,
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18, xaxis_rotate=20)

line = Line("", width=800)
line.add("", date, city_over_count.values+5, is_smooth=True)

overlap = Overlap(width=700)
overlap.add(bar)
overlap.add(line)

overlap

NameError: name 'city_over_count' is not defined